src/catalog/src/durable/impls/persist.rs

```diff
@@ -650,6 +652,48 @@ impl OpenableDurableCatalogState for UnopenedPersistCatalogState {
     }
 }
 
+/// Certain large collections are only needed during startup. This struct helps us cache these
+/// values during startup and ignore them at all other times.
+#[derive(Debug)]
+enum LargeCollectionStartupCache<T> {
+    /// Actively caching any new value that is seen.
+    Open(Vec<T>),
+    /// Ignoring any new value that is seen.
+    Closed,
+}
+
+impl<T> LargeCollectionStartupCache<T> {
+    /// Create a new opened [`LargeCollectionStartupCache`].
+    fn new_open() -> LargeCollectionStartupCache<T> {
+        LargeCollectionStartupCache::Open(Vec::new())
+    }
+
+    /// Add value to cache if open, otherwise ignore.
+    fn push(&mut self, value: T) {
+        match self {
+            LargeCollectionStartupCache::Open(cache) => {
+                cache.push(value);
+            }
+            LargeCollectionStartupCache::Closed => {
+                // Cache is closed, ignore value.
+            }
+        }
+    }
+
+    /// If the cache is open, then return all the cached values and close the cache, otherwise
+    /// return `None`.
+    fn take(&mut self) -> Option<Vec<T>> {
+        match self {
+            LargeCollectionStartupCache::Open(cache) => {
+                let cache = Some(std::mem::take(cache));
+                *self = LargeCollectionStartupCache::Closed;
+                cache
+            }
+            LargeCollectionStartupCache::Closed => None,
+        }
```

Comment:
You've basically inlined `mem::replace` here; it's a bit simpler if you just call it.

```suggestion
        if let Self::Open(cache) = std::mem::replace(self, Self::Closed) {
            Some(cache)
        } else {
            None
        }
```